### ISSUE TITLE
Upgrade error: Table 'ship.tabStudent Batch Student' doesn't exist

### ISSUE THREAD

**Issue by jevonearth** [frappe/erpnext#8875]
Here are two failing patches because the `"Table 'ship.tabStudent Batch Student' doesn't exist"`

```
Executing erpnext.patches.v8_0.merge_student_batch_and_student_group in ship (ship)
Traceback (most recent call last):
  File "/usr/lib/python2.7/runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/utils/bench_helper.py", line 79, in <module>
    main()
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/utils/bench_helper.py", line 16, in main
    click.Group(commands=commands)(prog_name='bench')
  File "/home/erpnext/frappe-bench/env/local/lib/python2.7/site-packages/click/core.py", line 700, in __call__
    return self.main(*args, **kwargs)
  File "/home/erpnext/frappe-bench/env/local/lib/python2.7/site-packages/click/core.py", line 680, in main
    rv = self.invoke(ctx)
  File "/home/erpnext/frappe-bench/env/local/lib/python2.7/site-packages/click/core.py", line 1027, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/erpnext/frappe-bench/env/local/lib/python2.7/site-packages/click/core.py", line 1027, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/erpnext/frappe-bench/env/local/lib/python2.7/site-packages/click/core.py", line 873, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/erpnext/frappe-bench/env/local/lib/python2.7/site-packages/click/core.py", line 508, in invoke
    return callback(*args, **kwargs)
  File "/home/erpnext/frappe-bench/env/local/lib/python2.7/site-packages/click/decorators.py", line 16, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/commands/__init__.py", line 24, in _func
    ret = f(frappe._dict(ctx.obj), *args, **kwargs)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/commands/site.py", line 214, in migrate
    migrate(context.verbose, rebuild_website=rebuild_website)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/migrate.py", line 31, in migrate
    frappe.modules.patch_handler.run_all()
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/modules/patch_handler.py", line 29, in run_all
    if not run_single(patchmodule = patch):
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/modules/patch_handler.py", line 63, in run_single
    return execute_patch(patchmodule, method, methodargs)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/modules/patch_handler.py", line 83, in execute_patch
    frappe.get_attr(patchmodule.split()[0] + ".execute")()
  File "/home/erpnext/frappe-bench/apps/erpnext/erpnext/patches/v8_0/merge_student_batch_and_student_group.py", line 15, in execute
    program, academic_year, academic_term from `tabStudent Batch`''', as_dict=1)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/database.py", line 149, in sql
    self._cursor.execute(query)
  File "/home/erpnext/frappe-bench/env/local/lib/python2.7/site-packages/MySQLdb/cursors.py", line 205, in execute
    self.errorhandler(self, exc, value)
  File "/home/erpnext/frappe-bench/env/local/lib/python2.7/site-packages/MySQLdb/connections.py", line 36, in defaulterrorhandler
    raise errorclass, errorvalue
_mysql_exceptions.ProgrammingError: (1146, "Table 'ship.tabStudent Batch' doesn't exist")
```

```
Executing erpnext.patches.v7_2.mark_students_active in ship (ship)
/home/erpnext/frappe-bench/apps/erpnext/erpnext/schools/doctype/student_batch_student/student_batch_student.json missing
Traceback (most recent call last):
  File "/usr/lib/python2.7/runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/utils/bench_helper.py", line 79, in <module>
    main()
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/utils/bench_helper.py", line 16, in main
    click.Group(commands=commands)(prog_name='bench')
  File "/home/erpnext/frappe-bench/env/local/lib/python2.7/site-packages/click/core.py", line 700, in __call__
    return self.main(*args, **kwargs)
  File "/home/erpnext/frappe-bench/env/local/lib/python2.7/site-packages/click/core.py", line 680, in main
    rv = self.invoke(ctx)
  File "/home/erpnext/frappe-bench/env/local/lib/python2.7/site-packages/click/core.py", line 1027, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/erpnext/frappe-bench/env/local/lib/python2.7/site-packages/click/core.py", line 1027, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/erpnext/frappe-bench/env/local/lib/python2.7/site-packages/click/core.py", line 873, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/erpnext/frappe-bench/env/local/lib/python2.7/site-packages/click/core.py", line 508, in invoke
    return callback(*args, **kwargs)
  File "/home/erpnext/frappe-bench/env/local/lib/python2.7/site-packages/click/decorators.py", line 16, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/commands/__init__.py", line 24, in _func
    ret = f(frappe._dict(ctx.obj), *args, **kwargs)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/commands/site.py", line 214, in migrate
    migrate(context.verbose, rebuild_website=rebuild_website)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/migrate.py", line 31, in migrate
    frappe.modules.patch_handler.run_all()
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/modules/patch_handler.py", line 29, in run_all
    if not run_single(patchmodule = patch):
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/modules/patch_handler.py", line 63, in run_single
    return execute_patch(patchmodule, method, methodargs)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/modules/patch_handler.py", line 83, in execute_patch
    frappe.get_attr(patchmodule.split()[0] + ".execute")()
  File "/home/erpnext/frappe-bench/apps/erpnext/erpnext/patches/v7_2/mark_students_active.py", line 6, in execute
    frappe.db.sql("update `tabStudent Batch Student` set active=1")
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/database.py", line 149, in sql
    self._cursor.execute(query)
  File "/home/erpnext/frappe-bench/env/local/lib/python2.7/site-packages/MySQLdb/cursors.py", line 205, in execute
    self.errorhandler(self, exc, value)
  File "/home/erpnext/frappe-bench/env/local/lib/python2.7/site-packages/MySQLdb/connections.py", line 36, in defaulterrorhandler
    raise errorclass, errorvalue
_mysql_exceptions.ProgrammingError: (1146, "Table 'ship.tabStudent Batch Student' doesn't exist")
```

**Comment by ajinkyagadewar** [comment]
Error is still coming:

Executing erpnext.patches.v7_2.update_assessment_modules in site1.local (1bd3e0294d)
Traceback (most recent call last):
  File "/usr/lib/python2.7/runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/home/erp/frappe-bench/apps/frappe/frappe/utils/bench_helper.py", line 79, in <module>
    main()
  File "/home/erp/frappe-bench/apps/frappe/frappe/utils/bench_helper.py", line 16, in main
    click.Group(commands=commands)(prog_name='bench')
  File "/home/erp/frappe-bench/env/local/lib/python2.7/site-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "/home/erp/frappe-bench/env/local/lib/python2.7/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/home/erp/frappe-bench/env/local/lib/python2.7/site-packages/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/erp/frappe-bench/env/local/lib/python2.7/site-packages/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/erp/frappe-bench/env/local/lib/python2.7/site-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/erp/frappe-bench/env/local/lib/python2.7/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/home/erp/frappe-bench/env/local/lib/python2.7/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/erp/frappe-bench/apps/frappe/frappe/commands/__init__.py", line 24, in _func
    ret = f(frappe._dict(ctx.obj), *args, **kwargs)
  File "/home/erp/frappe-bench/apps/frappe/frappe/commands/site.py", line 214, in migrate
    migrate(context.verbose, rebuild_website=rebuild_website)
  File "/home/erp/frappe-bench/apps/frappe/frappe/migrate.py", line 31, in migrate
    frappe.modules.patch_handler.run_all()
  File "/home/erp/frappe-bench/apps/frappe/frappe/modules/patch_handler.py", line 29, in run_all
    if not run_single(patchmodule = patch):
  File "/home/erp/frappe-bench/apps/frappe/frappe/modules/patch_handler.py", line 63, in run_single
    return execute_patch(patchmodule, method, methodargs)
  File "/home/erp/frappe-bench/apps/frappe/frappe/modules/patch_handler.py", line 83, in execute_patch
    frappe.get_attr(patchmodule.split()[0] + ".execute")()
  File "/home/erp/frappe-bench/apps/erpnext/erpnext/patches/v7_2/update_assessment_modules.py", line 6, in execute
    frappe.rename_doc("DocType", "Grading Structure", "Grading Scale", force=True)
  File "/home/erp/frappe-bench/apps/frappe/frappe/__init__.py", line 668, in rename_doc
    return rename_doc(*args, **kwargs)
  File "/home/erp/frappe-bench/apps/frappe/frappe/model/rename_doc.py", line 42, in rename_doc
    rename_parent_and_child(doctype, old, new, meta)
  File "/home/erp/frappe-bench/apps/frappe/frappe/model/rename_doc.py", line 99, in rename_parent_and_child
    (new, old))
  File "/home/erp/frappe-bench/apps/frappe/frappe/database.py", line 138, in sql
    self._cursor.execute(query, values)
  File "/home/erp/frappe-bench/env/local/lib/python2.7/site-packages/MySQLdb/cursors.py", line 205, in execute
    self.errorhandler(self, exc, value)
  File "/home/erp/frappe-bench/env/local/lib/python2.7/site-packages/MySQLdb/connections.py", line 36, in defaulterrorhandler
    raise errorclass, errorvalue
_mysql_exceptions.IntegrityError: (1062, "Duplicate entry 'Grading Scale' for key 'PRIMARY'")

**Comment by jevonearth** [comment]
Hi @manassolanki , Any progress on this one? I'm happy to help with testing/verification.

This issue is currently a blocker for us to upgrade.
Thanks,
-Jev

**Comment by nabinhait** [comment]
> _mysql_exceptions.ProgrammingError: (1146, "Table 'ship.tabStudent Batch Student' doesn't exist")
This issue has been fixed.

**Comment by nabinhait** [comment]
> _mysql_exceptions.IntegrityError: (1062, "Duplicate entry 'Grading Scale' for key 'PRIMARY'")

In the past, have you skipped the patch `/home/erp/frappe-bench/apps/erpnext/erpnext/patches/v7_2/update_assessment_modules.py`?

**Comment by jevonearth** [comment]
@nabinhait Can you cut a release so I can get this fix?

Yes, I may have commented some table modifications out in `/home/erp/frappe-bench/apps/erpnext/erpnext/patches/v7_2/update_assessment_modules.py` in order to get the update to complete and do some testing.

I will re-test as soon as your fixes get released. Thanks!

**Comment by jevonearth** [comment]
@nabinhait Just tested this, and I get the 
`_mysql_exceptions.IntegrityError: (1062, "Duplicate entry 'Grading Scale' for key 'PRIMARY'")` error WITHOUT having skipped anything in `/home/erp/frappe-bench/apps/erpnext/erpnext/patches/v7_2/update_assessment_modules.py`

**Comment by nabinhait** [comment]
Actually order of the patch is very important. If you skipped a patch, it can lead to many other issues.

I will try to modify the patch, will check if it fixable or not.

**Comment by jevonearth** [comment]
Okay, but I did not skipped the patch today. I just happened to be trying the update as you updated this ticket. I did not get the `_mysql_exceptions.ProgrammingError: (1146, "Table 'ship.tabStudent Batch Student' doesn't exist")` error, so I believe I got your hotfix,

I do still get the `_mysql_exceptions.IntegrityError: (1062, "Duplicate entry 'Grading Scale' for key 'PRIMARY'"`

I messaged you privately on gitter too, if you need any specific information you can message me there. Happy to get you the info you need :)

**Comment by jevonearth** [comment]
Today's testing was done with versions:
``` 
erpnext 8.0.36
frappe 8.0.57
```

**Comment by nabinhait** [comment]
Pushed a fix https://github.com/frappe/erpnext/pull/8974/files, it should fix the error.

It will be released tomorrow.

**Comment by jevonearth** [comment]
Super, thanks. I will import it into my test system today and test. All going well, I will be finally able to upgrade to v8 tomorrow! :)

**Comment by jevonearth** [comment]
I imported the patch manually yesterday and it worked great. I'm going to test in now from master. Did this fix get included in v8.0.38?
